torch/fx/experimental/Partitioner.py

```diff
@@ -393,3 +439,86 @@ def get_bfs_level_partition(self) -> None:
+           embedding node, a new partition is created for the next node.
+           After the partition, the partitions are combined as much as possible.
+           The rule is that non-embedding partition only combines non-embedding ones.
+           So as the embedding ones.
+        """
+        # Track memory size for embedding partitions and non-embedding partitions separately
+        embedding_partition_to_used_mem_bytes: Dict[Partition, int] = {}
+        non_embedding_partition_to_used_mem_bytes: Dict[Partition, int] = {}
+        # A Flag to check the boundary
+        partition_embeddings: bool = False
+        partition = self.create_partition()
+        used_mem_bytes = 0
+        for node in self.graph_module.graph.nodes:
+            if node.op in {'call_module', 'call_method', 'call_function'}:
+                # Check if crossing the boundary between embedding nodes and non embedding nodes
+                if (self.is_embedding_node(node) and (not partition_embeddings)) \
+                        or ((not self.is_embedding_node(node)) and partition_embeddings):
+                    # Crossing the boundary
+                    # Check if the current partition is an empty partition
+                    if used_mem_bytes != 0:
+                        # The current partition isn't an empty partition. Create a new one.
+                        if partition_embeddings:
+                            embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
+                        else:
+                            non_embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
+                        partition = self.create_partition()
+                        used_mem_bytes = 0
+                    partition_embeddings = not partition_embeddings
+                total_size_of_input_nodes = self.get_total_size_of_input_nodes(node, partition)
+                if used_mem_bytes + total_size_of_input_nodes > available_mem_bytes:
+                    # The current partition is full. Create a new one.
+                    if partition_embeddings:
+                        embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
+                    else:
+                        non_embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
+                    partition = self.create_partition()
+                    used_mem_bytes = 0
+                    if total_size_of_input_nodes > available_mem_bytes:
+                        raise RuntimeError(node.target + 'is too large to fit into a device')
+                partition.add_node(node)
+                used_mem_bytes = used_mem_bytes + total_size_of_input_nodes
+        if partition_embeddings:
+            embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
+        else:
+            non_embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
```

Comment:
The region from 470-504 can be made simpler and easier to read by factoring out the logic relating to completing a partition into a function:

```
        partition = None
        def reset_partition(new_partition=True):
            if partition is not None:
                # Create new partition
                if partition_embeddings:
                    embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
                else:
                    non_embedding_partition_to_used_mem_bytes[partition] = used_mem_bytes
            if new_partition:
                partition = self.create_partition()
                used_mem_bytes = 0
        for node in self.graph_module.graph.nodes:
            if node.op in {'call_module', 'call_method', 'call_function'}:
                # Check if crossing the boundary between embedding nodes and non embedding nodes
                if self.is_embedding_node(node) != partition_embeddings:
                    # Crossing the boundary
                    # Check if the current partition is an empty partition
                    if used_mem_bytes != 0:
                        # The current partition isn't an empty partition. Create a new one.
                        reset_partition()
                    partition_embeddings = not partition_embeddings
                total_size_of_input_nodes = self.get_total_size_of_input_nodes(node, partition)
                if used_mem_bytes + total_size_of_input_nodes > available_mem_bytes:
                    # The current partition is full. Create a new one.
                    reset_partition()
                    if total_size_of_input_nodes > available_mem_bytes:
                        raise RuntimeError(node.target + 'is too large to fit into a device')
                partition.add_node(node)
                used_mem_bytes = used_mem_bytes + total_size_of_input_nodes
        reset_partition(new_partition = False)
```